### PR TITLE
add support for Maildir flags

### PR DESCRIPTION
**Pull request by josuah**
Hello and thank you for pursuing the development of mdsort all this time, I am using it ever since in my mail server.

I did not find a way to add a Maildir flag to incoming email, so I tried to add something doing that.
Here is what I have, but while the mail inherit the flag appropriately, they do not get filed into any mailbox, and get dropped!

Here is the conf fragment I am using for adding a flag:

```
stdin {
        match   header { "To" } /@panoramix-labs.fr/i
        mdflags "F"

        [...]
}
```

Here is what I could observe with ktrace:

Getting renamed all right:
```
  9389 mdsort   CALL  renameat(3,0x9c398a66400,5,0x7f7ffffdd010)
  9389 mdsort   NAMI  "1661165408.9389_64.ams1"
  9389 mdsort   NAMI  "1661165408.9389_59.ams1:2,F"
```

But later:
```
  9389 mdsort   CALL  getdents(3,0x9c393ccc000,0x10000)
  9389 mdsort   RET   getdents 120/0x78
  9389 mdsort   CALL  kbind(0x7f7ffffdd0e8,24,0x8b3a0a59361cdc48)
  9389 mdsort   RET   kbind 0
  9389 mdsort   CALL  unlinkat(3,0x9c393ccc058,0<>)
  9389 mdsort   NAMI  "1661165408.9389_59.ams1:2,F"
  9389 mdsort   RET   unlinkat 0
  9389 mdsort   CALL  getdents(3,0x9c393ccc000,0x10000)
  9389 mdsort   RET   getdents 0
  9389 mdsort   CALL  kbind(0x7f7ffffdd0e8,24,0x8b3a0a59361cdc48)
  9389 mdsort   RET   kbind 0
  9389 mdsort   CALL  rmdir(0x9c398a5fbf0)
  9389 mdsort   NAMI  "/tmp//mdsort-druywni8/new"
  9389 mdsort   RET   rmdir 0
  9389 mdsort   CALL  rmdir(0x9c398a5f7f0)
  9389 mdsort   NAMI  "/tmp//mdsort-druywni8"
  9389 mdsort   RET   rmdir 0
```

And that is it, no moving to any other maildir, regardless of the options.

### CODE REVIEW

**Comment by josuah** [comment]
Do you have any idea of what could go wrong?

Adding the flag works with `maildir "~/Maildir/INBOX" { ... }`, but not with `stdin { ... }`. Maybe I missed something obvious, or in the evaluation mechanism.

**Comment by mptre** [comment]
In order to make your new action work with stdin, you will need an
additional move action. Otherwise, the message ends up being moved to a
new file inside the temporary maildir used during MDA delivery. This is
something that mdsort probably should detect and exit with a temporary
failure to ensure no emails are lost even when running with a
sub-optimal configuration. Another option is to make use of the pass
action if you have a following catch all rule.

	stdin {
	  match ... flags "F" move "~/Maildir/INBOX"
	}

	stdin {
	  match ... flags "F" pass
	  match all move "~/Maildir/INBOX"
	}

This happens to work while using maildir (i.e. not stdin) as the message
will implicitly be moved to the maildir it resides in.

File: parse.y [hunk 1]
```diff
@@ -565,13 +579,15 @@ yylex(void)
...
 		{ "discard",		DISCARD },
+		{ "draft",		DRAFT },
 		{ "exec",		EXEC },
 		{ "flag",		FLAG },
 		{ "header",		HEADER },
 		{ "isdirectory",	ISDIRECTORY },
 		{ "label",		LABEL },
 		{ "maildir",		MAILDIR },
 		{ "match",		MATCH },
+		{ "mdflags",		MDFLAGS },
```
Comment (mptre) [inline parse.y]: Please call the new action `flags` and make sure to update the expr_type and eval routine accordingly.

File: expr.c [hunk 2]
```diff
@@ -673,6 +678,22 @@ expr_eval_match(struct expr *ex, struct expr_eval_arg *ea)
...
+static int
+expr_eval_mdflags(struct expr *ex, struct expr_eval_arg *ea)
+{
+	struct match *mh;
+	const char *flags;
+
+	mh = match_alloc(ex, ea->ea_msg);
+	flags = TAILQ_FIRST(ex->ex_strings)->val;
+	for (; *flags != '\0'; flags++)
+		message_flags_set(&ea->ea_msg->me_mflags, *flags, 1);
```
Comment (mptre) [inline expr.c]: Check for errors.

File: mdsort.conf.5 [hunk 3]
```diff
@@ -327,6 +327,22 @@ to the X-Label header in message.
 The
 .Ar label
 is interpolated.
+.It Ic mdflags Dq Ar flags
```
Comment (mptre) [inline mdsort.conf.5]: I wouldn't enumerate the common flags here, mdsort accepts any flag within the [A-Za-z] range.

File: parse.y [hunk 4]
```diff
@@ -326,6 +328,18 @@ expraction	: BREAK {
 			strings_append(strings, $2);
 			expr_set_strings($$, strings);
 		}
+		| MDFLAGS STRING {
+			struct string_list *strings;
+			char *flags;
+
+			$$ = expr_alloc(EXPR_TYPE_MDFLAGS, lineno, NULL, NULL);
+			strings = strings_alloc();
+			flags = strdup($2);
```
Comment (mptre) [inline parse.y]: I don't think you should make a copy here, strings_append() takes ownership of the given string anyway?

Comment (josuah) [reply parse.y]: Shorter code, one allocation less, this is good news.

File: expr.c (hunk 2 above)
Comment (josuah) [reply expr.c]: Should it be err() in risk of loosing email?
I push -f'ed a warn() in the meantime.

File: mdsort.conf.5 (hunk 3 above)
Comment (josuah) [reply mdsort.conf.5]: That also makes the documentation shorter.

File: parse.y (hunk 1 above)
Comment (josuah) [reply parse.y]: "flags" sounds more natural than "mdflags" (4 consonants!)

**Comment by josuah** [comment]
Thank you for your review! My bug was hiding in the place I did expect it the least, but also the one I checked the less: the configuration!

**Comment by josuah** [comment]
With the new configuration file syntax it works smoothly!

Instead of having two separate rules (it did sound fishy to begin with) matching the same emails...

File: expr.c (hunk 2 above)
Comment (mptre) [reply expr.c]: You should return EXPR_ERROR instead since that will cause evaluation to be aborted and the error to trickle all the way up.

**Review by mptre — COMMENTED** [review]
This is starting to look good. Still some things I would like to address on my own before merging but that will hopefully be done later this week.

**Comment by mptre** [comment]
Merged, thanks. Assuming it looks good on your end, I will create a new release soon.

**Comment by josuah** [comment]
Yes looking good! I only forgot to remove the [WIP].
Thanks for the advises and streamlining this feature.

**Pull request closed without merging**